radicle-tracker/src/metadata/clock.rs

```diff
@@ -0,0 +1,263 @@
+//!
+//! # Examples
+//! ```
+//! # use std::error::Error;
+//! #
+//! # fn main() -> Result<(), Box<dyn Error>> {
+//! use radicle_tracker::clock::{Clock, Elapsed, RadicleClock};
+//! use std::thread::sleep;
+//! use std::time::Duration;
+//!
+//! let now = RadicleClock::current_time();
+//! sleep(Duration::new(1, 0));
+//! let then = RadicleClock::current_time();
+//!
+//! let elapsed = now.elapsed(&then)?;
+//!
+//! assert_eq!(elapsed, Elapsed::Minutes(0));
+//! #
+//! #     Ok(())
+//! # }
+//! ```
+
+// Rough calculations for the number of seconds in some larger unit
+const SECONDS_IN_MINUTE: u64 = 60;
+const SECONDS_IN_HOUR: u64 = SECONDS_IN_MINUTE * 60;
+const SECONDS_IN_DAY: u64 = SECONDS_IN_HOUR * 24;
+const SECONDS_IN_WEEK: u64 = SECONDS_IN_DAY * 7;
+const SECONDS_IN_MONTH: u64 = SECONDS_IN_WEEK * 4;
+const SECONDS_IN_YEAR: u64 = SECONDS_IN_MONTH * 12;
+
+use std::time::{Duration, SystemTime, SystemTimeError};
+
+/// The elapsed time from some previous moment in the past. This is to capture
+/// concepts like, "this comment was posted 4 minutes ago", or "5 days ago", or
+/// "1 year ago".
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum Elapsed {
+    /// Elapsed "x minutes ago".
+    Minutes(u64),
+    /// Elapsed "x hours ago".
+    Hours(u64),
+    /// Elapsed "x days ago".
+    Days(u64),
+    /// Elapsed "x weeks ago".
+    Weeks(u64),
+    /// Elapsed "x months ago".
+    Months(u64),
+    /// Elapsed "x years ago".
+    Years(u64),
+}
+
+impl std::fmt::Display for Elapsed {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Elapsed::Minutes(m) => write!(f, "{} minutes ago", m),
+            Elapsed::Hours(m) => write!(f, "{} hours ago", m),
+            Elapsed::Days(m) => write!(f, "{} days ago", m),
+            Elapsed::Weeks(m) => write!(f, "{} weeks ago", m),
+            Elapsed::Months(m) => write!(f, "{} months ago", m),
+            Elapsed::Years(m) => write!(f, "{} years ago", m),
+        }
+    }
+}
+
+/// `Clock` captures the minimal functionality for grabbing a point in time and
+/// calculating the duration since another point in time.
+pub trait Clock {
+    /// Time calculations tend to return errors.
+    type Error;
+
+    /// Get the current timestamp.
+    fn current_time() -> Self;
+
+    /// Get the duration for a timestamp in the past, `self`, up to the point of
+    /// `other`. It should return [`std::time::Duration`], which can be used
+    /// for further calculation.
+    fn duration_since(&self, other: &Self) -> Result<Duration, Self::Error>;
+}
+
+/// A minimal clock for getting a moment in time and calculating the duration up
+/// to another moment in time.
+///
+/// It does this through its instance of [`Clock`] trait.
+///
+/// For usability, it also provides a function [`RadicleClock::elapsed`] which
+/// returns an [`Elapsed`] value for presentation purposes.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
```

Comment:
Might be a good idea to explain why there's no `PartialOrd` instance, so that someone doesn't in the future add it, thinking you forgot it :wink: